src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -3456,49 +3457,58 @@ and build_let_rec_env _top_level env lbs : list<letbinding> * env_t * guard_t =
      else
        None
    in
+   let check_annot univ_vars t =
+       let env0 = Env.push_univ_vars env0 univ_vars in
+       let t, _, g = tc_check_tot_or_gtot_term ({env0 with check_uvars=true}) t (fst <| U.type_u()) "" in
+       env0, g |> Rel.resolve_implicits env |> Rel.discharge_guard env0, t
```

Comment:
This code is only hoisted as a separate function, the text looks same otherwise?

Independently, just curious would anything go wrong if we did not `resolve_implicits` and `discharge_guard` here? And return the guard as is?

Reply:
This is an interesting point. I spent a while thinking about this and I decided to keep it as is for this PR. But, would like to improve it separately.

Currently, the behavior of not re-typechecking the lbtyp in certain situations is really very fragile and potentially dubious, or even unsound. But I have yet to find a way to trigger it. I preserved the behavior of the existing code in this PR, but it would be better to be stricter about it. 

The interaction with 2-phase typechecking in particular is problematic and has been like this in master already.

Consider what happens when you have

```
let rec f (x:t) : C = e
```

After the first phase is complete, it becomes

```
let rec f : (x:t -> C) = fun (x:t) -> (e <: C)
```

i.e., the lbtyp is set on `f`

Then, the second phase considers that the lbtyp is an annotated type of `f` that came from a val declaration and then proceeds to skip checking it for well-formedness. In this case, I think it still gets trapped because the ascription remains. But, it's pretty dodgy, since the ascription is checked in a slightly different context (extended with the let rec bindings).

Reply:
Leaving this conversation open for further post-merge discussion